File: kbs_bbs/bbs2www/src/bbsfwd.c
#include "bbslib.h"

int main() {
	struct fileheader *x;
	char board[80], file[80], target[80];
	struct userec *u = NULL;

	init_all();
	strsncpy(board, getparm("board"), 30);
	strsncpy(file, getparm("file"), 30);
	strsncpy(target, getparm("target"), 30);
	if(!loginok) http_fatal("匆匆过客不能进行本项操作");
	if(!has_read_perm(currentuser, board)) http_fatal("错误的讨论区");
	x=get_file_ent(board, file);
	if(x==0) http_fatal("错误的文件名");
	printf("<center>%s -- 转寄/推荐给好友 [使用者: %s]<hr color=green>\n", BBSNAME, currentuser->userid);
	if(target[0]) {
		if(!strstr(target, "@")) {
			if(getuser(target, &u) == 0)
				http_fatal("错误的使用者帐号");
			strcpy(target, u->userid);
		}
		return do_fwd(x, board, target);
	}
	printf("<table><tr><td>\n");
	printf("文章标题: %s<br>\n", nohtml(x->title));
	printf("文章作者: %s<br>\n", x->owner);
	printf("原讨论区: %s<br>\n", board);
	printf("<form action=bbsfwd method=post>\n");
	printf("<input type=hidden name=board value=%s>", board);
	printf("<input type=hidden name=file value=%s>", file);
	printf("把文章转寄给 <input name=target size=30 maxlength=30 value=%s> (请输入对方的id或email地址). <br>\n",
		currentuser->email);
	printf("<input type=submit value=确定转寄></form>");
	return 0;
}

int do_fwd(struct fileheader *x, char *board, char *target) {
	FILE *fp, *fp2;
	char title[512], buf[512], path[200], i;
	sprintf(path, "boards/%s/%s", board, x->filename);
	if(!file_exist(path)) http_fatal("文件内容已丢失, 无法转寄");
	sprintf(title, "[转寄] %s", x->title);
	title[60]=0;
	post_mail(target, title, path, currentuser->userid, currentuser->username, fromhost, -1);
	printf("文章已转寄给'%s'<br>\n", nohtml(target));
	printf("[<a href='javascript:history.go(-2)'>返回</a>]");
}

File: kbs_bbs/bbs2www/html/bbsfwd.php
<?php
	require("funcs.php");
	if ($loginok != 1)
		html_nologin();
	else
	{
		html_init("gb2312");
        if ($currentuser["userid"]=="guest")
            html_error_quit("匆匆过客不能转寄文章");

		if( !isset($_GET["board"]) && !isset($_POST["board"]))
			html_error_quit("错误的讨论区");
		if( isset($_GET["board"]) )
			$board = $_GET["board"];
		else
			$board = $_POST["board"];

		$brdarr = array();
		$brdnum = bbs_getboard($board, $brdarr);
		if ($brdnum == 0){
			html_error_quit("错误的讨论区1");
		}
		$usernum = $currentuser["index"];
		if (bbs_checkreadperm($usernum, $brdnum) == 0){
			html_error_quit("错误的讨论区2");
		}

		if( !isset($_GET["id"]) && !isset($_POST["id"]))
			html_error_quit("错误的文章号");
		if( isset($_GET["id"]) )
			$id = $_GET["id"];
		else
			$id = $_POST["id"];

		$articles = array ();
		$num = bbs_get_records_from_id($brdarr["NAME"], $id, 
			$dir_modes["NORMAL"] , $articles);
		if($num == 0)
			html_error_quit("错误的文章号");
?>
<center>
<?php echo BBS_FULL_NAME; ?> -- 转寄文章 [使用者: <?php echo $currentuser["userid"];?>]
<hr color="green"><br>
</center>
<?php
		if ( !isset($_POST["submit"])){
?>
文章标题: <?php echo htmlspecialchars($articles[1]["TITLE"]);?><br>
文章作者: <?php echo $articles[1]["OWNER"];?><br>
原讨论区: <?php echo $brdarr["NAME"];?><br>
<form action="/bbsfwd.php" method="post">
<input type="hidden" name="board" value="<?php echo $brdarr["NAME"];?>">
<input type="hidden" name="id" value="<?php echo $id;?>">
把文章转寄给 <input type="text" name="target" size="40" maxlength="69" value="<?php echo $currentuser["email"];?>"> (请输入对方的id或email地址). <br>
<input type="checkbox" name="big5" value="1"> 使用BIG5码<br>
<input type="checkbox" name="noansi" value="1" checked> 过滤ANSI控制符<br>
<input type="submit" name="submit" value="确定转寄">
</form>
<?php
			html_normal_quit();
		}

		if( isset($_POST["target"]) )
			$target =  $_POST["target"];
		else
			$target = "";
		if($target == "")
			html_error_quit("请指定对象");

		if( isset($_POST["big5"]) )
			$big5 = $_POST["big5"];
		else
			$big5=0;
		settype($big5, "integer");

		if( isset($_POST["noansi"]) )
			$noansi = $_POST["noansi"];
		else
			$noansi=0;
		settype($noansi, "integer");

		$ret = bbs_doforward($brdarr["NAME"], $articles[1]["FILENAME"], $articles[1]["TITLE"], $target, $big5, $noansi);
		if($ret < 0)
			html_error_quit("系统错误:".$ret);
?>
文章已转寄给'<?php echo $target;?>'<br>
[<a href="javascript:history.go(-2)">返回</a>]
<?php
	}
?>
